torch/cuda/amp/amp_scaler.py

```diff
@@ -0,0 +1,292 @@
+import torch
+from collections import defaultdict
+from torch._six import container_abcs
+
+
+class AmpScaler(object):
+    """
+    Performs dynamic gradient scaling, intended to be used when gradients
+    have a dtype with limited dynamic range (e.g. if part of your network is ``torch.float16``).
+
+    Here's how that looks in a simple example::
+
+        scaler = AmpScaler()
+
+        for input, target in data:
+            optimizer.zero_grad()
+            output = model(input)
+            loss = loss_fn(output, target)
+            scaler.scale(loss).backward()
+            scaler.step(optimizer)
+            scaler.update()
+
+    See :ref:`Gradient Scaling Examples<gradient-scaling-examples>` for usage in more complex cases.
+
+    By default, ``scaler.step`` internally unscales ``optimizer``'s gradients before applying them, so the
+    learning rate and other hyperparameters don't need to change.  If you wish to unscale gradients manually
+    prior to :meth:`step`, use :meth:`unscale`.
+
+    ``scaler`` maintains the scale factor internally.  To leverage the dtype's full dynamic range,
+    ``scaler`` attempts to use the largest scale factor it can without incurring overflow.
+    It does so by checking the gradients for infs and NaNs during every :meth:`step` or separate :meth:`unscale`.
+    If no infs/NaNs are found, :meth:`step` runs the underlying ``optimizer.step()`` as usual and
+    :meth:`update` increases the scale factor slightly.  If infs/NaNs are found, :meth:`step` skips the underlying
```

Comment:
Not "increases the scale factor slightly." It multiplies the scale factor by the growth factor.